src/include/storage/checkpoint_io.h

```diff
@@ -0,0 +1,254 @@
+   * @param row_buffer pointer to the ProjectedRow buffer.
+   * @param layout of the Row.
+   */
+  void SerializeTuple(ProjectedColumns::RowView *row, ProjectedRow *row_buffer, const storage::BlockLayout &layout);
+
+  /**
+   * Serialize a ProjectedRow and its varlens into the file.
+   * @param row_buffer pointer to the row to be serialized.
+   * @param total_varlen total length of all varlens of this row.
+   * @param varlen_entries all the varlen entries in this row.
+   */
+  void AppendTupleToBuffer(ProjectedRow *row_buffer, int32_t total_varlen,
+                           const std::vector<const VarlenEntry *> &varlen_entries);
+
+  /**
+   * Get the current checkpoint page.
+   * @return pointer to the checkpoint page header.
+   */
+  CheckpointFilePage *GetPage() { return reinterpret_cast<CheckpointFilePage *>(buffer_); }
+
+ private:
+  int out_;  // fd of the output files
+  uint32_t block_size_;
+  uint32_t cur_buffer_size_ = 0;
+  byte *buffer_ = nullptr;
+
+  void ResetBuffer() {
+    memset(buffer_, 0, block_size_);
+    CheckpointFilePage::Initialize(reinterpret_cast<CheckpointFilePage *>(buffer_));
+    cur_buffer_size_ = sizeof(CheckpointFilePage);
+  }
+
+  void PersistBuffer() {
+    // TODO(zhaozhe): calculate CHECKSUM. Currently using default 0 as checksum
+    if (cur_buffer_size_ == sizeof(CheckpointFilePage)) {
+      // If the buffer has no contents, just return
+      return;
+    }
+    PosixIoWrappers::WriteFully(out_, buffer_, block_size_);
+    ResetBuffer();
+  }
+};
+
+/**
+ * Reader class to read checkpoint file in pages (Checkpoint Page),
+ * and provides basic interface similar to an iterator.
+ */
+class BufferedTupleReader {
+ public:
+  /**
+   * Instantiate a new BufferedTupleReader to read from a checkpoint file.
+   * @param log_file_path path to the checkpoint file.
+   */
+  explicit BufferedTupleReader(const char *log_file_path)
+      : in_(PosixIoWrappers::Open(log_file_path, O_RDONLY)), buffer_(new byte[block_size_]()) {}
+
+  /**
+   * destruct the reader, close the file opened and release the buffer space.
+   */
+  ~BufferedTupleReader() {
+    PosixIoWrappers::Close(in_);
+    delete[] buffer_;
+  }
+
+  /**
+   * Read the next block from the file into the buffer.
+   * @return true if a page is read successfully. false, if EOF is encountered.
+   */
+  bool ReadNextBlock() {
+    uint32_t size = PosixIoWrappers::ReadFully(in_, buffer_, block_size_);
+    if (size == 0) {
+      return false;
+    }
+    TERRIER_ASSERT(size == block_size_, "Incomplete Checkpoint Page");
+    page_offset_ += static_cast<uint32_t>(sizeof(CheckpointFilePage));
+    return true;
+  }
+
+  /**
+   * Read the next row from the buffer.
+   * @return pointer to the row if the next row is available.
+   *         nullptr if an error happens or there is no other row in the page.
+   */
+  ProjectedRow *ReadNextRow() {
+    if (block_size_ - page_offset_ < sizeof(uint32_t)) {
```

Comment:
I don't see any resetting page_offset_. So what happens in the case you have already called ReadNextBlock() a few times and you are on some n'th block? In that case wouldn't page_offset_ be greater than the block_size so this check would always be true even if that block has data in it.

Reply:
This is correct.